compiler/rustc_mir_transform/src/coverage/counters.rs

```diff
@@ -496,4 +500,126 @@ impl<'a> CountersBuilder<'a> {
 
         None
     }
+
+    fn into_coverage_counters(self) -> CoverageCounters {
+        Transcriber::new(&self).transcribe_counters()
+    }
+}
+
+/// Helper struct for converting `CountersBuilder` into a final `CoverageCounters`.
+struct Transcriber<'a> {
+    old: &'a CountersBuilder<'a>,
+    new: CoverageCounters,
+    phys_counter_for_site: FxHashMap<Site, BcbCounter>,
+}
+
+impl<'a> Transcriber<'a> {
+    fn new(old: &'a CountersBuilder<'a>) -> Self {
+        Self {
+            old,
+            new: CoverageCounters::with_num_bcbs(old.graph.num_nodes()),
+            phys_counter_for_site: FxHashMap::default(),
+        }
+    }
+
+    fn transcribe_counters(mut self) -> CoverageCounters {
```

Comment:
So this is basically a normalization if there are no old counters and a system for keeping the diff small if there are old counters?

Reply:
This is kind of tricky to explain, but it's also the whole crux of this PR, so I should try.

We can think of this code as going through a series of refactoring stages:

- Graph traversal → `CoverageCounters` *(status quo)*
- Graph traversal → `CoverageCounters` → `Transcriber` → simplified `CoverageCounters`
- Graph traversal → `FxHashMap<Site, SiteCounter>` → `Transcriber` → simplified `CoverageCounters`

The main goal of introducing `Transcriber` as a middle layer is so that the part before `Transcriber` can be changed to not be tied to `CoverageCounters`. To make that feasible, we need to go through the intermediate step of having two different `CoverageCounters` (*old* and *new*), so that we can then replace the first one with something else.

Reply:
The fact that final `CoverageCounters` is simpler than the original one starts off as being a bonus extra, but it also lets the earlier steps not care so much about producing “optimal” results in a single pass. I expect that to be a big help in future changes to how counter creation works.

Reply:
That makes sense, thanks!